File: 2021/07/15/longest_common_prefix.py
"""
https://leetcode.com/problems/longest-common-prefix/
Author: Andrew Jarombek
Date: 7/15/2021
"""

# Stay strong, you are so loved.  Keep my love close to you & let it help you through whatever challenges you face.
# Racing a 2 mile on the track this evening in CT with Joe.

from typing import List


def longest_common_prefix(strs: List[str]) -> str:
    prefix = strs[0]

    for i in range(1, len(strs)):
        current_str = strs[i]
        for j in range(len(prefix)):
            if current_str[:j + 1] != prefix[:j + 1]:
                prefix = current_str[:j]
                break

    return prefix


if __name__ == '__main__':
    input1 = ["flower", "flow", "flight"]
    assert longest_common_prefix(input1) == "fl"

    input2 = ["dog", "racecar", "car"]
    assert longest_common_prefix(input2) == ""

File: 2021/09/03/c/longestCommonPrefix.c
// Program: https://leetcode.com/problems/longest-common-prefix/
// Author: Andrew Jarombek
// Date: 9/3/2021

#include <string.h>
#include <assert.h>
#include <stdlib.h>

char* longestCommonPrefix(char** strs, int strsSize) {
    char* commonPrefix = strs[0];

    for (int i = 1; i < strsSize; i++) {
        char* current = strs[i];
        int currentLength = strlen(current);
        int prefixLength = strlen(commonPrefix);

        if (currentLength < prefixLength) {
            prefixLength = currentLength;
            commonPrefix[prefixLength] = '\0';
        } else if (prefixLength < currentLength) {
            current[prefixLength] = '\0';
        }

        while (strcmp(commonPrefix, current) != 0) {
            int newLength = strlen(commonPrefix) - 1;

            if (newLength == 0) {
                return "";
            }

            commonPrefix[newLength] = '\0';
            current[newLength] = '\0';
        }
    }

    return commonPrefix;
}

int main() {
    char* first = "flower";
    char* second = "flow";
    char* third = "flight";

    char** str = (char**) calloc(3, sizeof(char*));
    str[0] = first;
    str[1] = second;
    str[2] = third;

    char* result = longestCommonPrefix(str, 3);
    assert(strcmp(result, "fl") == 0);

    return 0;
}
